llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

```diff
@@ -9115,14 +9115,45 @@ bool SIInstrInfo::isAsmOnlyOpcode(int MCOp) const {
   }
 }
 
+#define GENERATE_RENAMED_GFX9_CASES(OPCODE)                                    \
+  case OPCODE##_dpp:                                                           \
+  case OPCODE##_e32:                                                           \
+  case OPCODE##_e64:                                                           \
+  case OPCODE##_e64_dpp:                                                       \
+  case OPCODE##_sdwa:
+
+static bool isRenamedInGFX9(int Opcode) {
+  switch (Opcode) {
+    GENERATE_RENAMED_GFX9_CASES(AMDGPU::V_ADDC_U32)
+    GENERATE_RENAMED_GFX9_CASES(AMDGPU::V_ADD_CO_U32)
+    GENERATE_RENAMED_GFX9_CASES(AMDGPU::V_ADD_U32)
+    GENERATE_RENAMED_GFX9_CASES(AMDGPU::V_SUBBREV_U32)
+    GENERATE_RENAMED_GFX9_CASES(AMDGPU::V_SUBB_U32)
+    GENERATE_RENAMED_GFX9_CASES(AMDGPU::V_SUBREV_CO_U32)
+    GENERATE_RENAMED_GFX9_CASES(AMDGPU::V_SUBREV_U32)
+    GENERATE_RENAMED_GFX9_CASES(AMDGPU::V_SUB_CO_U32)
+    GENERATE_RENAMED_GFX9_CASES(AMDGPU::V_SUB_U32)
+  //
+  case AMDGPU::V_DIV_FIXUP_F16_gfx9_e64:
+  case AMDGPU::V_FMA_F16_gfx9_e64:
+  case AMDGPU::V_INTERP_P2_F16:
+  case AMDGPU::V_MAD_F16_e64:
+  case AMDGPU::V_MAD_U16_e64:
+  case AMDGPU::V_MAD_I16_e64:
+    return true;
+  default:
+    return false;
+  }
+}
+
 int SIInstrInfo::pseudoToMCOpcode(int Opcode) const {
   Opcode = SIInstrInfo::getNonSoftWaitcntOpcode(Opcode);
 
   unsigned Gen = subtargetEncodingFamily(ST);
 
-  if ((get(Opcode).TSFlags & SIInstrFlags::renamedInGFX9) != 0 &&
-    ST.getGeneration() == AMDGPUSubtarget::GFX9)
+  if (isRenamedInGFX9(Opcode) && ST.getGeneration() == AMDGPUSubtarget::GFX9) {
```

Comment:
```suggestion
  if (ST.getGeneration() == AMDGPUSubtarget::GFX9 && isRenamedInGFX9(Opcode)) {
```

Reply:
But this moves knowledge from tablegen, into manually written functions

Reply:
I agree that's a downside, but a minor one IMO:

- This information is needed in a single place (still a single source of truth, it's just no longer TG)
- This affects a small amount of instructions (no need to automatically generate it)
- We're unlikely to add any new instructions in that category (so changing C++ isn't a bigger annoyance than changing TG)

In the end it's really a matter of taste, I personally like seeing our TableGen get simpler with minimal downsides, but I'm also okay if everyone else prefers to keep as much as possible in TableGen

Reply:
Once we remove the bit from SIInstrFormats my thought was to have all the information together in the C++ file since it seems like the only place it's really needed/used. Don't have a strong opinion.

Reply:
@arsenm do you feel strongly here? If so, what is the better alternative in your view, move the isRenamedInGFX9 back into the TableGen files?

Reply:
> I'm not sure I ever understood this bit in the first place; why does the rename have to come up in the encoding emission? Can we avoid needing to know this in the first place here?

It's because GFX8 and GFX9 use the same Real for some instructions. The obvious way to support this would be to have it directly represented in the `getMCOpcodeGen` table, i.e. have the same value (Real) in multiple columns (EncodingFamilies) of the same row (Pseudo). But the table is defined with `InstrMapping` which does not let you do that.

Reply:
> Do we really have a pressing need to scavenge a bit right now?

I'm not sure it's pressing but @kzhuravl has an on going effect to reduce the number of TableGen bits we use and this was decided upon as one of the lower hanging opportunities.

Reply:
I'm confused, InstrMapping allows you to define arbitrary columns? I think we should fix that if there's an issue

Reply:
I'm confused too. I'm no longer confident that what I said about `InstrMapping` is true. I've been meaning to take another look.